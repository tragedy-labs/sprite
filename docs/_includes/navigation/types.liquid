{%
  include navigation/list.liquid
  title="Types"
  items=site.types
  icon="icons/ts.svg"
%}
