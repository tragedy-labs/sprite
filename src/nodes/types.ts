export { ArcadeDeleteReturnCount } from './delete/returnCount.js';
export { ArcadeResultSortDirection } from './select/from/direction.js';
export { ArcadeSelectTimeoutStrategy } from './select/from/strategy.js';
