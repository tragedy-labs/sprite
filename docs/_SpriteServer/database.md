---

layout: default

title: database

permalink: /SpriteServer/database.html

---

### _SpriteServer_.database

#### Interface

(**databaseName: *string***)

Returns an `SpriteDatabase` instance for the supplied `databaseName`,
using the authorization details of the `SpriteServer` instance.

#### Example

```ts
const server = new SpriteServer({
  username: 'aUser',
  password: 'aPassword',
  address: 'http://localhost:2480',
});

async function databaseExample() {
  try {
    const database = await server.database('aDatabase');
    // returns an instance of SpriteDatabase
    console.log(database.name);
  } catch (error) {
    // handle errors
    console.error(error);
  }
};

databaseExample();
```

