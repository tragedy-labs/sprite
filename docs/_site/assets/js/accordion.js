/*
 *   This content is licensed according to the W3C Software License at
 *   https://www.w3.org/copyright/software-license-2023/
 *
 *   Simple accordion pattern example
 */

'use strict';

class Accordion {
  constructor(domNode) {
    this.root = domNode;
    this.button = this.root.querySelector('button[aria-expanded]');
    this.icon = this.root.querySelector('svg');
    const controlsId = this.button.getAttribute('aria-controls');
    this.content = document.getElementById(controlsId);

    this.open = this.button.getAttribute('aria-expanded') === 'true';

    // add event listeners
    this.button.addEventListener('click', this.onButtonClick.bind(this));
    this.toggle(false);
  }

  onButtonClick() {
    this.toggle(!this.open);
  }

  toggle(open) {
    // don't do anything if the open state doesn't change
    if (open === this.open) {
      return;
    }

    // update the internal state
    this.open = open;

    // handle DOM updates
    this.button.setAttribute('aria-expanded', `${open}`);
    if (open) {
      this.content.removeAttribute('hidden');
      this.icon.classList.add('open');
    } else {
      this.content.setAttribute('hidden', '');
      this.icon.classList.remove('open');
    }
  }

  // Add public open and close methods for convenience
  open() {
    this.toggle(true);
  }

  close() {
    this.toggle(false);
  }
}

const initializeAccordions = () => {
  // init accordions
  const accordions = document.querySelectorAll('.menuItem');
  accordions.forEach((accordionEl) => {
    new Accordion(accordionEl);
  });
};

document.addEventListener('DOMContentLoaded', initializeAccordions);
