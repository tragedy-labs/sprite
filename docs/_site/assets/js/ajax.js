// Constants
const CONTENT_CONTAINER_ID = "content";
const PARSER = new DOMParser();

// State
let historyStack = [];
let currentStateIndex = -1;

// Functions
async function fetchDocument(url) {
  try {
    const response = await fetch(url);
    return await response.text();
  } catch (error) {
    throw new Error(`Failed to fetch document: ${url}`, { cause: error });
  }
}

async function parseHtml(html) {
  return PARSER.parseFromString(html, "text/html");
}

async function updateContent(htmlDoc) {
  const title = htmlDoc.querySelector("title").textContent;
  const contentDiv = htmlDoc.getElementById(CONTENT_CONTAINER_ID);
  const target = document.getElementById(CONTENT_CONTAINER_ID);
  target.innerHTML = contentDiv.innerHTML;
  target.scrollTop = 0;
  document.title = title;

  // Reinitialize link click events for newly loaded content
  createLinkClickEvents(target);
}

function pushToHistory(url) {
  historyStack = historyStack.slice(0, currentStateIndex + 1);
  historyStack.push(url);
  currentStateIndex++;
  history.pushState({ index: currentStateIndex }, "", url);
}

async function navigate(url, addToHistory = true) {
  const urlObj = new URL(url, window.location.origin);

  if (urlObj)
    if (urlObj.hash && urlObj.pathname === window.location.pathname) {
      // Handle local page anchor navigation
      const anchor = document.querySelector(urlObj.hash);
      if (anchor) {
        anchor.scrollIntoView();
        addToHistory && pushToHistory(url);
        return;
      }
    }

  // Handle Local Navigation
  const html = await fetchDocument(url);
  const htmlDoc = await parseHtml(html);
  await updateContent(htmlDoc);

  addToHistory && pushToHistory(url);
}

function createLinkClickEvents(root) {
  const links = root.querySelectorAll("a");
  links.forEach((link) => {
    link.addEventListener("click", (e) => {
      const url = e.currentTarget.href;
      if (isLocalLink(url)) {
        e.preventDefault();
        navigate(url);
      }
    });
  });
}

function isLocalLink(url) {
  const linkUrl = new URL(url, window.location.origin);
  return linkUrl.origin === window.location.origin;
}

window.addEventListener("popstate", (e) => {
  if (e.state && e.state.index !== undefined) {
    currentStateIndex = e.state.index;
    navigate(historyStack[currentStateIndex], false);
  }
});

// Initialize
function initAjax() {
  createLinkClickEvents(document);
  // Initialize the first state
  historyStack.push(window.location.href);
  currentStateIndex++;
  history.replaceState({ index: currentStateIndex }, "", window.location.href);
}

// Call init
document.addEventListener("DOMContentLoaded", initAjax);
