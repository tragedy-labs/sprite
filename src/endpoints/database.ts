export const endpoints = {
  query: '/api/v1/query',
  command: '/api/v1/command',
  beginTransaction: '/api/v1/begin',
  commitTransaction: '/api/v1/commit',
  rollbackTransaction: '/api/v1/rollback'
} as const;
