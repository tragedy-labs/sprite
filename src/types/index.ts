export * from './client.js';
export * from './commands.js';
export * from './database.js';
export * from './edge.js';
export * from './operators.js';
export * from './queries.js';
export * from './server.js';
export * from './transaction.js';
