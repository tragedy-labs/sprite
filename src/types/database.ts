import { ArcadeSelectTimeoutStrategy } from '../nodes/types.js';
import { ISpriteRestClientConnectionParameters } from './client.js';
import { SpriteRestClient } from '../SpriteRestClient.js';
import { SpriteTransaction } from '../SpriteTransaction.js';
import { ValidSuperTypeKey } from './type.js';

/**
 * The possible values for a parameter of a command or query
 */
export type SpriteAllowedParamValues = string | boolean | number;

/**
 * The possible categories of a record in ArcadeDB,
 * as they appear in the record's `@cat` property.
 */
export type ArcadeRecordCategory = 'd' | 'e' | 'v';

/**
 * The possible record types, as expected to be
 * supplied to a `CREATE TYPE` sql operation.
 */
export type ArcadeRecordType = 'document' | 'edge' | 'vertex';

type RecordMetaBase<C extends ArcadeRecordCategory> = {
  /** The id of the record. */
  '@rid': string;
  /** The category of the record. */
  '@cat': C;
  /** The type of the record. */
  '@type': string;
};

/**
 * The base metadata for record in
 * ArcadeDB.
 */
export type RecordMeta<C extends ArcadeRecordCategory = ArcadeRecordCategory> =
  C extends 'e' ? EdgeRecordMeta : RecordMetaBase<C>;

/**
 * The metadata for an Edge record in
 * ArcadeDB
 */
export interface EdgeRecordMeta extends RecordMetaBase<'e'> {
  /** The `@rid` of the vertex the edge goes to */
  '@in': string;
  /** The `@rid` of the vertex the edge comes from */
  '@out': string;
}

/**
 * Adds ArcadeDB record metadata to the types
 * defined in the schema.
 */
export type AsArcadeEdges<S> = {
  [K in keyof S]: S[K] & EdgeRecordMeta;
};

/**
 * Adds ArcadeDB record metadata to the types
 * defined in the schema.
 */
export type AsArcadeRecords<S, C extends ArcadeRecordCategory> = {
  [K in keyof S]: S[K] & RecordMeta<C>;
};

export type AsArcadeDocuments<S> = AsArcadeRecords<S, 'd'>;
export type AsArcadeVertices<S> = AsArcadeRecords<S, 'v'>;

/**
 * The TypeNames in a supplied schema.
 */
export type TypeNames<Schema> = keyof Schema;

/**
 * The parameters necessary to perform opertations
 * on a target ArcadeDB database.
 */
export interface ISpriteDatabaseConnectionParameters
  extends ISpriteRestClientConnectionParameters {
  /**
   * The name of the database to connect to.
   */
  databaseName: string;
}

/**
 * The parameters used to create a `SpriteDatabase`
 * intance when there is an existing instance of a
 * `SpriteRestClient`, such as when intantiated
 * from a method in the `SpriteServer` class
 */
export interface ISpriteDatabaseClientParameters {
  /**
   * The `SpriteRestClient` instance
   */
  client: SpriteRestClient;
  /**
   * The name of the database to connect to.
   */
  databaseName: string;
}

/**
 * The Query languages supported by ArcadeDB, supplied as a parameter
 * to `SpriteDatabase.query()` and `SpriteDatabase.command()`
 */
export type ArcadeSupportedQueryLanguages =
  | 'sql'
  | 'sqlscript'
  | 'graphql'
  | 'cypher'
  | 'gremlin'
  | 'mongo';

/**
 * Options to create a new type with.
 */
export interface ISpriteCreateTypeOptions<S, N extends TypeNames<S>> {
  /**
   * When set to true, the type creation will be ignored if the
   * type already exists (instead of failing with an error).
   * @default false
   */
  ifNotExists?: boolean;
  /**
   * Defines a super-type you want to extend with this type.
   */
  extends?: ValidSuperTypeKey<S, N>;
  /**
   * A bucket-name, or an array of bucket-names you want this type to use.
   */
  buckets?: string | Array<string>;
  /**
   * Defines the total number of buckets you want to create for this type. The
   * @default 1
   */
  totalBuckets?: number;
}

export type OmitMeta<T> = Omit<T, keyof RecordMeta | keyof EdgeRecordMeta>;

/** Options to insert a new record with */
export interface ISpriteInsertRecordOptions<T> {
  /** The data to populate the newly created reacord with */
  data?: OmitMeta<T> | OmitMeta<T>[];
  /** The bucket to store the record in */
  bucket?: string;
}

/**
 * Options for a database.dropType() command
 */
export interface ISpriteDropTypeOptions {
  /**
   * Prevent errors if the type does not exits when attempting to drop it.
   * @default false;
   */
  ifExists?: boolean;
  /**
   * Defines whether the command drops non-empty edge and vertex types. Note, this can
   * disrupt data consistency. Be sure to create a backup before running it.
   * @default: false
   */
  unsafe?: boolean;
}

/** A callback which contains the operations to be conducted within a transaction */
export type SpriteTransactionCallback =
  | ((trx: SpriteTransaction) => Promise<void>)
  | ((trx: SpriteTransaction) => void);

/**
 * How the bucket will be selected for a newly created record of this type
 * @default 'round-robin'
 */
export type ArcadeBucketSelectionStrategies =
  | 'round-robin'
  | 'thread'
  | 'partitioned<primary-key>';

/**
 * A type definition returned by ArcadeDB when a getSchema command
 * is compeleted.
 */
export type ArcadeTypeDefinition = {
  /** The name of the type */
  name: string;
  /** The category of the type (document, vertex, or edge) */
  type: ArcadeRecordType;
  /**
   * The number of records with this type name
   */
  records: number;
  /**
   * The buckets associated with this type.
   */
  buckets: Array<string>;
  /** How the bucket will be selected for a newly created record of this type */
  bucketSelectionStrategy: ArcadeBucketSelectionStrategies;
  /**
   * Super types associated with this type.
   */
  parentTypes: Array<string>;
  /**
   * Properties defined for this type.
   */
  properties: Array<string>;
  /** Indexes defined for records of this type */
  indexes: Array<string>;
  /** An object for custom user-defined properties for this type */
  custom: Record<string, unknown>;
};

/** The response from a `SpriteDatabase.getSchema()` query */
export type ArcadeGetSchemaResponse = ArcadeTypeDefinition[];

export type ArcadeSqlExaplantionExecutionPlanStep = {
  cost: number;
  description: string;
  javaType: string;
  name: string;
  subSteps: ArcadeSqlExaplantionExecutionPlanStep[];
  targetNode: string;
  type: string;
};

/** SQL Explanation Execution Plan */
export type ArcadeSqlExplanationExecutionPlan = {
  type: 'QueryExecutionPlan' | 'DDLExecutionPlan';
  javaType: string;
  cost: number;
  prettyPrint: string;
  steps: ArcadeSqlExaplantionExecutionPlanStep[];
};

/**
 * An object representing the explanation of an
 * SQL command.
 */
export type ArcadeSqlExplanation = {
  executionPlan: ArcadeSqlExplanationExecutionPlan;
  executionPlanAsString: string;
};

/**
 * The object returned from the server following a successfully command
 */
export interface ArcadeCommandResponse<T = unknown> {
  /**
   * The user that issued the command / query.
   */
  user: string;
  /** The version of ArcadeDB returning the result. */
  version: string;
  /**
   * The name of the server returning the result.
   */
  serverName: string;
  /**
   * The result of the command / query.
   */
  result: T;
}

/**
 * The object returned from the server following a successfully query
 */
export type ArcadeQueryResponse<T = unknown> = ArcadeCommandResponse<T>;

// TODO: have not checked these for compatibility with ArcadeDB
export declare const COMPARISON_OPERATORS: readonly [
  '=',
  '==',
  '!=',
  '<>',
  '>',
  '>=',
  '<',
  '<=',
  'in',
  'not in',
  'is',
  'is not',
  'like',
  'not like',
  'match',
  'ilike',
  'not ilike',
  '@>',
  '<@',
  '&&',
  '?',
  '?&',
  '!<',
  '!>',
  '<=>',
  '!~',
  '~',
  '~*',
  '!~*',
  '@@',
  '@@@',
  '!!',
  '<->',
  'regexp',
  'is distinct from',
  'is not distinct from'
];

/** Operators for a `WHERE` sql statement */
export type SpriteOperators = (typeof COMPARISON_OPERATORS)[number];

/** Options for a `selectFrom` operation, as executed via a `SpriteDatabase` modality. */
export interface ISpriteSelectFromOptions<
  S,
  N extends keyof S,
  P extends keyof WithRid<S, N>
> {
  /**
   * Designates conditions to filter the result-set.
   */
  where?: SpriteWhereClause<S, N, P>;
  /**
   * Designates the field with which to order the result-set.
   * Use the optional 'ASC' and 'DESC' operators to define the direction of the order.
   */
  orderBy?: {
    field: keyof S[N];
    /**
     * Defines the direction to sort the result (ASCending or DESCending).
     * @default 'ASC'.
     */
    direction?: 'ASC' | 'DESC';
  };
  /**
   * Defines the number of records you want to skip from the start of the result-set.
   * You mayfind this useful in Pagination, when using it in conjunction with the
   * limit `option`.
   */
  skip?: number;
  /**
   * Defines the maximum number of records in the result-set. You may find this useful in
   * Pagination, when using it in conjunction with the `skip` option.
   */
  limit?: number;
  /**
   * Defines the maximum time in milliseconds for the query, and optionally the
   * exception strategy to use.
   */
  timeout?: {
    /**
     *  The duration of the timeout in milliseconds.
     */
    duration: number;
    /**
     * The timeout strategy to use.\
     * `RETURN` Truncates the result-set, returning the data collected up to the timeout.\
     * `EXCEPTION` Raises an exception.
     */
    strategy?: ArcadeSelectTimeoutStrategy;
  };
}

/**
 * A base type, with an `@rid` property added.
 */
export type WithRid<S, N extends TypeNames<S>> = OmitMeta<S[N]> & {
  '@rid': string;
};

/**
 * An array with three items used to describe a `where` statement.
 * @example ['@rid', '==', '#0:0']
 */
export type SpriteWhereClause<
  S,
  N extends TypeNames<S>,
  P extends keyof WithRid<S, N>
> = [P, SpriteOperators, WithRid<S, N>[P]];

/**
 * Options for a `deleteFrom` operation as executed
 * via a `SpriteDatabase` modality.
 */
export interface ISpriteDeleteFromOptions<
  S,
  N extends TypeNames<S>,
  P extends keyof WithRid<S, N>
> {
  /**
   * Designates conditions to filter the result-set.
   */
  where: SpriteWhereClause<S, N, P>;
  /**
   * The duration of the timeout in milliseconds.
   */
  timeout?: number;
  /**
   * Defines the maximum number of records in the result-set.
   * @default undefined
   */
  limit?: number;
  /**
   * Defines what is returned following the command: the count of the records before (`BEFORE`) or following deletion (`COUNT`).
   * @default 'COUNT'
   */
  return?: 'COUNT' | 'BEFORE';
}
