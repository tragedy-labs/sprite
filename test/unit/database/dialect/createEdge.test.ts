import { client } from './testClient.js';
import { endpoints } from '../../../../src/endpoints/database.js';
import {
  variables,
  headersWithTransaction as headers
} from '../../../variables.js';
import { EdgeTypes, VertexTypes } from '../../types.js';
import { testTransaction } from '../client/testClient.js';

const vertexName = 'aVertex';
const propertyName = 'aProperty';
const typeName = 'anEdge';
const SpriteDatabase = client.database;
type TypeName = typeof typeName;

const data = {
  aProperty: 'aValue'
};

const stringifiedData = JSON.stringify(data);

const indexDescriptor = {
  type: 'aVertex',
  key: 'aProperty',
  value: 'aValue'
} as const;

const indexSelectStatement = `(SELECT FROM ${vertexName} WHERE ${propertyName} = 'aValue')`;

const createEdgeTyped = client.createEdge<
  EdgeTypes,
  VertexTypes,
  TypeName,
  'aVertex',
  'aVertex'
>;

const createEdgeResult = {
  user: variables.username,
  version: '',
  serverName: '',
  result: [{ count: 0 }]
};

describe('SqlDialect.insertRecord()', () => {
  it(`should make a properly formatted POST request to ${endpoints.command}/${variables.databaseName}`, async () => {
    // Arrange
    jest.spyOn(global, 'fetch').mockResolvedValueOnce({
      status: 200,
      json: async () => createEdgeResult
    } as Response);

    // Act
    await createEdgeTyped(
      typeName,
      variables.rid,
      variables.rid,
      testTransaction,
      {
        data
      }
    );

    // Assert

    expect(global.fetch).toHaveBeenCalledWith(
      `${variables.address}${endpoints.command}/${variables.databaseName}`,
      {
        method: 'POST',
        headers,
        body: JSON.stringify({
          language: 'sql',
          command: `CREATE EDGE ${typeName} FROM ${variables.rid} TO ${variables.rid} CONTENT ${stringifiedData}`
        })
      }
    );
  });

  it(`handles string rids in the to and from fields by appending TO ${variables.rid} FROM ${variables.rid} to the command when supplied with ${variables.rid} in the TO and FROM fields`, async () => {
    // Arrange
    jest
      .spyOn(SpriteDatabase, 'command')
      .mockImplementationOnce(async () => createEdgeResult);

    // Act
    await createEdgeTyped(
      typeName,
      variables.rid,
      variables.rid,
      testTransaction
    );

    // Assert
    expect(SpriteDatabase.command).toHaveBeenCalledWith(
      'sql',
      `CREATE EDGE ${typeName} FROM ${variables.rid} TO ${variables.rid}`,
      testTransaction
    );
  });

  it('handles "ifNotExists" option by appending "IF NOT EXISTS" to the command when passed "true"', async () => {
    // Arrange
    jest.spyOn(SpriteDatabase, 'command').mockImplementationOnce(async () => {
      return createEdgeResult.result;
    });

    // Act
    await createEdgeTyped(
      typeName,
      variables.rid,
      variables.rid,
      testTransaction,
      { ifNotExists: true }
    );

    // Assert
    expect(SpriteDatabase.command).toHaveBeenCalledWith(
      'sql',
      `CREATE EDGE ${typeName} FROM ${variables.rid} TO ${variables.rid} IF NOT EXISTS`,
      testTransaction
    );
  });

  it(`handles index descriptor objects in the TO and FROM arguments option by appending TO ${indexSelectStatement} FROM ${indexSelectStatement} to the command when an index descriptor object is supplied"`, async () => {
    // Arrange
    jest
      .spyOn(SpriteDatabase, 'command')
      .mockImplementationOnce(async () => createEdgeResult);

    // Act
    await createEdgeTyped(
      typeName,
      indexDescriptor,
      indexDescriptor,
      testTransaction
    );

    // Assert
    expect(SpriteDatabase.command).toHaveBeenCalledWith(
      'sql',
      `CREATE EDGE ${typeName} FROM ${indexSelectStatement} TO ${indexSelectStatement}`,
      testTransaction
    );
  });

  it(`handles "data" option by appending CONTENT ${stringifiedData} to the command when data is set to ${data}`, async () => {
    // Arrange
    jest
      .spyOn(SpriteDatabase, 'command')
      .mockImplementationOnce(async () => createEdgeResult);

    // Act
    await createEdgeTyped(
      typeName,
      variables.rid,
      variables.rid,
      testTransaction,
      { data }
    );

    // Assert
    expect(SpriteDatabase.command).toHaveBeenCalledWith(
      'sql',
      `CREATE EDGE ${typeName} FROM ${variables.rid} TO ${variables.rid} CONTENT ${stringifiedData}`,
      testTransaction
    );
  });

  it(`handles "upsert" option by appending "UPSERT" to the command when upsert is set to true`, async () => {
    // Arrange
    jest
      .spyOn(SpriteDatabase, 'command')
      .mockImplementationOnce(async () => createEdgeResult);

    // Act
    await createEdgeTyped(
      typeName,
      variables.rid,
      variables.rid,
      testTransaction,
      {
        upsert: true
      }
    );

    // Assert
    expect(SpriteDatabase.command).toHaveBeenCalledWith(
      'sql',
      `CREATE EDGE ${typeName} UPSERT FROM ${variables.rid} TO ${variables.rid}`,
      testTransaction
    );
  });

  it(`handles "unidirectional" option by appending "UNIDIRECTIONAL" to the command when unidirectional is set to true`, async () => {
    // Arrange
    jest
      .spyOn(SpriteDatabase, 'command')
      .mockImplementationOnce(async () => createEdgeResult);

    // Act
    await createEdgeTyped(
      typeName,
      variables.rid,
      variables.rid,
      testTransaction,
      {
        unidirectional: true
      }
    );
    expect(SpriteDatabase.command).toHaveBeenCalledWith(
      'sql',
      `CREATE EDGE ${typeName} FROM ${variables.rid} TO ${variables.rid} UNIDIRECTIONAL`,
      testTransaction
    );
  });

  it(`handles "retry" option by appending "RETRY 4 WAIT 1000" to the command when retry.attempts is set to 4 and retry.wait is set to 1000`, async () => {
    // Arrange
    jest
      .spyOn(SpriteDatabase, 'command')
      .mockImplementationOnce(async () => createEdgeResult);

    // Act
    await createEdgeTyped(
      typeName,
      variables.rid,
      variables.rid,
      testTransaction,
      {
        retry: { attempts: 4, wait: 1000 }
      }
    );

    // Assert
    expect(SpriteDatabase.command).toHaveBeenCalledWith(
      'sql',
      `CREATE EDGE ${typeName} FROM ${variables.rid} TO ${variables.rid} RETRY 4 WAIT 1000`,
      testTransaction
    );
  });

  it(`handles "batchSize" option by appending "BATCH 1000" to the command when batchSize is set to 1000`, async () => {
    // Arrange
    jest
      .spyOn(SpriteDatabase, 'command')
      .mockImplementationOnce(async () => createEdgeResult);

    // Act
    await createEdgeTyped(
      typeName,
      variables.rid,
      variables.rid,
      testTransaction,
      {
        batchSize: 1000
      }
    );

    // Assert
    expect(SpriteDatabase.command).toHaveBeenCalledWith(
      'sql',
      `CREATE EDGE ${typeName} FROM ${variables.rid} TO ${variables.rid} BATCH 1000`,
      testTransaction
    );
  });

  it('propagates errors from internal methods', async () => {
    // Arrange
    jest
      .spyOn(SpriteDatabase, 'command')
      .mockRejectedValueOnce(new TypeError('Failed to fetch'));

    // Act
    await expect(
      createEdgeTyped(typeName, variables.rid, variables.rid, testTransaction)
    ).rejects.toMatchSnapshot();
  });
});
