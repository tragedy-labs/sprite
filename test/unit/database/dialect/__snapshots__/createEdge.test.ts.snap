// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SqlDialect.insertRecord() propagates errors from internal methods 1`] = `[Error: Unable to build CREATE EDGE sql command.]`;
